In [59]:
import pandas as pd

In [60]:
# read CSV file from the 'data' subdirectory using a relative path
df = pd.read_csv('data/age_structure_1997.csv')

# display the first 5 rows
df.head()

/home/asitha/anaconda3/lib/python3.7/site-packages/IPython/core/interactiveshell.py:3057: DtypeWarning: Columns (6) have mixed types. Specify dtype option on import or set low_memory=False.
  interactivity=interactivity, compiler=compiler, result=result)


YEAR  MONTH  SECTOR  DISTRICT  PSU_NO  HH_NO BLANK  RES_1 RES_2 RES_3  ...  \
0  1997      1       1         1      37      1     1      1              ...   
1  1997      1       1         1      37      1     2      1              ...   
2  1997      1       1         1      37      1     3      1              ...   
3  1997      1       1         1      37      1     4      1              ...   
4  1997      1       1         1      37      1     5      1              ...   

  Q37_LOOKING_FOR_WORK_IN_EACH Q38_THIS_PERSON_WAS Q39_MAIN_OCCUPATION  \
0                                                1                   1   
1                                                3                   2   
2                                                3                   2   
3                                                3                   2   
4                                                                        

  Q40A_MULTIPLY_WAGE_DAYS_WORKED Q40B_ANY_PAYMENTS_IN_KIND  \
0                           8000                         1   
1                                                            
2                                                            
3                                                            
4                                                            

  Q41_EMPLOYER_THE_SECONDARY_ACTIV Q42A_EARNING_FROM_SECOND_JOB  \
0                                2                                
1                                2                                
2                                2                                
3                                2                                
4                                                                 

  Q42B_ANY_PAYMENTS_IN_KIND  INFLATION_FACTOR  AGE_CATEGORY  
0                                 234393.4436             2  
1                                 234393.4436             2  
2                                 234393.4436             1  
3                                 234393.4436             1  
4                                 234393.4436             1  

[5 rows x 89 columns]

In [61]:
# display the last 5 rows
df.tail()

YEAR  MONTH  SECTOR  DISTRICT  PSU_NO  HH_NO BLANK  RES_1 RES_2 RES_3  \
80012  1997     12       2        25     131     76     2      1               
80013  1997     12       2        25     131     76     3      1               
80014  1997     12       2        25     131     76     4      1               
80015  1997     12       2        25     131     76     5      1               
80016  1997     12       2        25     131     76     6      1               

       ... Q37_LOOKING_FOR_WORK_IN_EACH Q38_THIS_PERSON_WAS  \
80012  ...                                                1   
80013  ...                                                3   
80014  ...                                                3   
80015  ...                                                    
80016  ...                                                    

      Q39_MAIN_OCCUPATION Q40A_MULTIPLY_WAGE_DAYS_WORKED  \
80012                   1                           1500   
80013                   2                                  
80014                   2                                  
80015                                                      
80016                                                      

      Q40B_ANY_PAYMENTS_IN_KIND Q41_EMPLOYER_THE_SECONDARY_ACTIV  \
80012                         2                                2   
80013                                                          2   
80014                                                          2   
80015                                                              
80016                                                              

      Q42A_EARNING_FROM_SECOND_JOB Q42B_ANY_PAYMENTS_IN_KIND  \
80012                                                          
80013                                                          
80014                                                          
80015                                                          
80016                                                          

       INFLATION_FACTOR  AGE_CATEGORY  
80012       234393.4436             2  
80013       234393.4436             2  
80014       234393.4436             1  
80015       234393.4436             1  
80016       234393.4436             1  

[5 rows x 89 columns]

In [62]:
# check the shape of the DataFrame (rows, columns)
df.shape

(80017, 89)

### Data Preprocessing

#### 1. Replcaing empty cells with NaN(Not a number)

In [63]:
import numpy as np
df.replace(' ', np.nan, inplace=True)
df.replace('  ', np.nan, inplace=True)
df.replace('   ', np.nan, inplace=True)
df.replace('    ', np.nan, inplace=True)
df.replace('     ', np.nan, inplace=True)
df.head()

YEAR  MONTH  SECTOR  DISTRICT  PSU_NO  HH_NO BLANK  RES_1 RES_2 RES_3  ...  \
0  1997      1       1         1      37      1     1      1   NaN   NaN  ...   
1  1997      1       1         1      37      1     2      1   NaN   NaN  ...   
2  1997      1       1         1      37      1     3      1   NaN   NaN  ...   
3  1997      1       1         1      37      1     4      1   NaN   NaN  ...   
4  1997      1       1         1      37      1     5      1   NaN   NaN  ...   

  Q37_LOOKING_FOR_WORK_IN_EACH Q38_THIS_PERSON_WAS Q39_MAIN_OCCUPATION  \
0                          NaN                   1                   1   
1                          NaN                   3                   2   
2                          NaN                   3                   2   
3                          NaN                   3                   2   
4                          NaN                 NaN                 NaN   

  Q40A_MULTIPLY_WAGE_DAYS_WORKED Q40B_ANY_PAYMENTS_IN_KIND  \
0                           8000                         1   
1                            NaN                       NaN   
2                            NaN                       NaN   
3                            NaN                       NaN   
4                            NaN                       NaN   

  Q41_EMPLOYER_THE_SECONDARY_ACTIV Q42A_EARNING_FROM_SECOND_JOB  \
0                                2                          NaN   
1                                2                          NaN   
2                                2                          NaN   
3                                2                          NaN   
4                              NaN                          NaN   

  Q42B_ANY_PAYMENTS_IN_KIND  INFLATION_FACTOR  AGE_CATEGORY  
0                       NaN       234393.4436             2  
1                       NaN       234393.4436             2  
2                       NaN       234393.4436             1  
3                       NaN       234393.4436             1  
4                       NaN       234393.4436             1  

[5 rows x 89 columns]

#### 2. Drop columns with NaN values 

In [64]:
df = df.dropna(axis=1, how='all')

#### NOTE: Best way to handle missing data

**Eliminating and Imputing** are two different strategies for handling samples or features with missing values.

Eliminating simply we remove the corresponding features (columns) or samples (rows) from the dataset entirely that have a certain number of missing values. The disadvantages for this strategy, we will run the risk of losing valuable information that our classifier needs to discriminate between classes.

Imputing we handle missing data by applying different interpolation techniques to estimate the missing values. One of the most common interpolation techniques is mean imputation, where we simply replace the missing value by the mean value of the entire feature column. A convenient way to achieve this is by using the Imputer class from scikit-learn, but filling missing values with this way is also a prediction which ignore the relation between missing values and other features.

Visualizing the important characteristics of a dataset is an important and recommended first step to clarifying the correlation between features to guide us to use a regression or another simple model to predict the values of missing variables. We have tried imputation here because we cannot loose data records.

#### 2. Imputation of missing values

##### i. Univariate vs. Multivariate Imputation

One type of imputation algorithm is univariate, which imputes values in the i-th feature dimension using only non-missing values in that feature dimension (e.g. impute.SimpleImputer). By contrast, multivariate imputation algorithms use the entire set of available feature dimensions to estimate the missing values (e.g. impute.IterativeImputer). As we can see we have to try multivariate imputation because it is the most sophisticated approach is to use. (***IterativeImputer class***)

In [65]:
from sklearn.impute import SimpleImputer
imp=SimpleImputer(missing_values=np.NaN)
idf=pd.DataFrame(imp.fit_transform(df))
idf.columns=df.columns
idf.index=df.index

In [66]:
idf.head()

YEAR  MONTH  SECTOR  DISTRICT  PSU_NO  HH_NO  BLANK  RES_1     RES_2  \
0  1997.0    1.0     1.0       1.0    37.0    1.0    1.0    1.0  5.190058   
1  1997.0    1.0     1.0       1.0    37.0    1.0    2.0    1.0  5.190058   
2  1997.0    1.0     1.0       1.0    37.0    1.0    3.0    1.0  5.190058   
3  1997.0    1.0     1.0       1.0    37.0    1.0    4.0    1.0  5.190058   
4  1997.0    1.0     1.0       1.0    37.0    1.0    5.0    1.0  5.190058   

      RES_3  ...  Q37_LOOKING_FOR_WORK_IN_EACH  Q38_THIS_PERSON_WAS  \
0  1.176316  ...                     43.430132             1.000000   
1  1.176316  ...                     43.430132             3.000000   
2  1.176316  ...                     43.430132             3.000000   
3  1.176316  ...                     43.430132             3.000000   
4  1.176316  ...                     43.430132             2.123759   

   Q39_MAIN_OCCUPATION  Q40A_MULTIPLY_WAGE_DAYS_WORKED  \
0             1.000000                     8000.000000   
1             2.000000                     3289.362064   
2             2.000000                     3289.362064   
3             2.000000                     3289.362064   
4             1.747177                     3289.362064   

   Q40B_ANY_PAYMENTS_IN_KIND  Q41_EMPLOYER_THE_SECONDARY_ACTIV  \
0                   1.000000                          2.000000   
1                   1.760422                          2.000000   
2                   1.760422                          2.000000   
3                   1.760422                          2.000000   
4                   1.760422                          1.987811   

   Q42A_EARNING_FROM_SECOND_JOB  Q42B_ANY_PAYMENTS_IN_KIND  INFLATION_FACTOR  \
0                   1211.929635                       1.71       234393.4436   
1                   1211.929635                       1.71       234393.4436   
2                   1211.929635                       1.71       234393.4436   
3                   1211.929635                       1.71       234393.4436   
4                   1211.929635                       1.71       234393.4436   

   AGE_CATEGORY  
0           2.0  
1           2.0  
2           1.0  
3           1.0  
4           1.0  

[5 rows x 83 columns]

In [67]:
X = idf.drop(['INFLATION_FACTOR'], axis=1)
y = idf['INFLATION_FACTOR']

# Feature selection

#### Plotting original data with regression line

In [ ]:
import seaborn as sns
%matplotlib inline

# visualize the relationship between the features and the response using scatterplots
for i in idf.columns:
    sns.pairplot(idf, x_vars=i, y_vars='INFLATION_FACTOR', height=5, aspect=0.9, kind='reg')

/home/asitha/anaconda3/lib/python3.7/site-packages/seaborn/axisgrid.py:1270: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).
  squeeze=False)


Feature selection can be done in multiple ways but there are broadly 3 categories of it.
1. Filter Method
2. Wrapper Method
3. Embedded Method

(NOTE: 2 and 3 method are suited when lesser features (~20). Since there are more than 20 features we will use filter method.)

### 1. Filter Method
Plot the Pearson correlation heatmap and see the correlation of independent variables with the output variable. We will only select features which has correlation of above 0.5.

#### Pearson correlation heatmap

In [ ]:
import matplotlib.pyplot as plt
%matplotlib inline
#Using Pearson Correlation
plt.figure(figsize=(100,100))
cor = idf.corr()
sns.heatmap(cor, annot=True, cmap=plt.cm.Reds)
plt.show()

#### Getting the selected features using correltion matrix

In [ ]:
#Correlation with output variable
cor_target = abs(cor["INFLATION_FACTOR"])#Selecting highly correlated features
selected_features_FM = cor_target[cor_target>=0.5]
selected_features_FM = list(selected_features_FM.keys())
selected_features_FM

### 2. Wrapper Method
Feed the features to the selected Machine Learning algorithm and based on the model performance you add/remove the features. This is an iterative and computationally expensive process but it is more accurate than the filter method.

There are different wrapper methods such as Backward Elimination, Forward Selection, Bidirectional Elimination and RFE.

#### i. Backward Elimination
Check the performance of the model and then iteratively remove the worst performing features one by one till the overall performance of the model comes in acceptable range.

The performance metric used here to evaluate feature performance is pvalue. If the pvalue is above 0.05 then we remove the feature, else we keep it.

Here we are using OLS model which stands for “Ordinary Least Squares”. This model is used for performing linear regression.

In [ ]:
import statsmodels.api as sm
#Backward Elimination
cols = list(X.columns)
pmax = 1
while (len(cols)>0):
    p= []
    X_1 = X[cols]
    X_1 = sm.add_constant(X_1)
    model = sm.OLS(y,X_1).fit()
    p = pd.Series(model.pvalues.values[0:],index = cols)      
    pmax = max(p)
    feature_with_p_max = p.idxmax()
    if(pmax>0.05):
        cols.remove(feature_with_p_max)
    else:
        break
selected_features_BE = cols
print(selected_features_BE)

### NOTE: Since there are more than 20 feature we have used filter method. But filter method does not works with this dataset as we expected. So we have used backward eleimintaion. 

What are the features?
- **MONTH:**
- **SECTOR:**
- **DISTRICT:**
- **HH_NO:**
- **BLANK:** etc.

What is the response?
- **INFLATION_FACTOR**

What else do we know?
- Because the response variable is continuous, this is a **regression** problem.
- There are 80017 **observations** (represented by the rows), and each observation is a single labour.

In [ ]:
# use the list to select a subset of the original DataFrame
research_features = ['YEAR', 'DISTRICT', 'P4_SEX', 'AGE_CATEGORY']
X = idf[research_features]

# print the first 5 rows
X.head()

In [29]:
# check the type and shape of X
print(type(X))
print(X.shape)

<class 'pandas.core.frame.DataFrame'>
(80017, 9)


In [30]:
# select a Series from the DataFrame
y = idf['INFLATION_FACTOR']

# print the first 5 values
y.head()

YEAR
1997    237.8926
1997    237.8926
1997    237.8926
1997    237.8926
1997    237.8926
Name: INFLATION_FACTOR, dtype: float64

In [31]:
# check the type and shape of y
print(type(y))
print(y.shape)

<class 'pandas.core.series.Series'>
(80017,)


## Splitting X and y into training and testing sets

In [32]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=1,test_size=0.2)

In [33]:
# default split is 80% for training and 20% for testing
print(X_train.shape)
print(y_train.shape)
print(X_test.shape)
print(y_test.shape)

(64013, 9)
(64013,)
(16004, 9)
(16004,)


## Scaling Data

In [34]:
from sklearn.preprocessing import StandardScaler

scaler = StandardScaler()
scaler.fit(X_train)
X_train = scaler.transform(X_train)
X_test = scaler.transform(X_test)

## Finding the best linear regression algorithm

In [35]:
#import sklearn regression models
from sklearn.linear_model import LinearRegression, Ridge, Lasso, ElasticNet, Lars, OrthogonalMatchingPursuit, BayesianRidge, ARDRegression, SGDRegressor, PassiveAggressiveRegressor, RANSACRegressor, TheilSenRegressor, HuberRegressor
from sklearn.tree import DecisionTreeRegressor
from sklearn.gaussian_process import GaussianProcessRegressor
from sklearn.neural_network import MLPRegressor
from sklearn.neighbors import KNeighborsRegressor, RadiusNeighborsRegressor
from sklearn.svm import SVR, NuSVR, LinearSVR
from sklearn.kernel_ridge import KernelRidge
from sklearn.ensemble import RandomForestRegressor

#Dictionary of all models. All models intialized with no args. Can modify any of them to test various args.
models = {'LinearRegression': LinearRegression(),
          'Ridge': Ridge(),
          'Lasso': Lasso(),
          'ElasticNet': ElasticNet(),
          'Lars': Lars(),
          'OrthogonalMatchingPursuit': OrthogonalMatchingPursuit(),
          'BayesianRidge': BayesianRidge(),
          'ARDRegression': ARDRegression(),
          'SGDRegressor': SGDRegressor(),
          'PassiveAggressiveRegressor': PassiveAggressiveRegressor(),
          'RANSACRegressor': RANSACRegressor(),
          'TheilSenRegressor': TheilSenRegressor(),
          'HuberRegressor': HuberRegressor(),
          'DecisionTreeRegressor': DecisionTreeRegressor(),
          'GaussianProcessRegressor': GaussianProcessRegressor(),
          'MLPRegressor': MLPRegressor(),
          'KNeighborsRegressor': KNeighborsRegressor(),
          'RadiusNeighborsRegressor': RadiusNeighborsRegressor(),
          'SVR': SVR(gamma='scale'),
          'NuSVR': NuSVR(gamma='scale'),
          'LinearSVR': LinearSVR(),
          'KernelRidge': KernelRidge(),
          'RandomForestRegressor': RandomForestRegressor(n_estimators=500, oob_score=True, random_state=100)
         }

### NOTE: Non-linear Regressors
#### Decision Trees
Decision Trees, also referred to as Classification and Regression Trees (CART), work for both categorical and continuous input and output variables. It works by splitting the data into two or more homogeneous sets based on the most significant splitter among the independent variables. The best differentiator is the one that minimizes the cost metric. The cost metrics for a classification tree is often the entropy or the gini index, whereas, for a regression tree, the default metric is the mean squared error.
#### Random Forest
Decision Trees are useful, but the problem is that they often tend to overfit the training data leading to high variances in the test data. Random Forest algorithms overcome this shortcoming by reducing the variance of the decision trees. They are called 'Forest' because they are the collection, or ensemble, of several decision trees. One major difference between a Decision Tree and a Random Forest model is on how the splits happen. In Random Forest, instead of trying splits on all the features, a sample of features is selected for each split, thereby reducing the variance of the model

## Model evaluation metrics for regression

Evaluation metrics for classification problems, such as **accuracy**, are not useful for regression problems. Instead, we need evaluation metrics designed for comparing continuous values.

Let's create some example numeric predictions, and calculate **three common evaluation metrics** for regression problems:

**Mean Absolute Error** (MAE) is the mean of the absolute value of the errors:

$$\frac 1n\sum_{i=1}^n|y_i-\hat{y}_i|$$

**Mean Squared Error** (MSE) is the mean of the squared errors:

$$\frac 1n\sum_{i=1}^n(y_i-\hat{y}_i)^2$$

**Root Mean Squared Error** (RMSE) is the square root of the mean of the squared errors:

$$\sqrt{\frac 1n\sum_{i=1}^n(y_i-\hat{y}_i)^2}$$

Comparing these metrics:

- **MAE** is the easiest to understand, because it's the average error.
- **MSE** is more popular than MAE, because MSE "punishes" larger errors.
- **RMSE** is even more popular than MSE, because RMSE is interpretable in the "y" units.

In [36]:
#Import non-sklearn packages
import numpy as np
import time
from sklearn.metrics import mean_squared_error, r2_score
from sklearn.model_selection import cross_val_score

print_coef = False

for key, model in models.items():
    begin = time.time()
    model.fit(X_train,y_train)
    print(key + ' Train time: ' + str((time.time() - begin)/60) + " minutes")
    preds = model.predict(X_test)
    mse = mean_squared_error(y_test,preds)
    r2 = r2_score(y_test,preds)
    scores = cross_val_score(model, X_train, y_train, cv=5)
    print(key + ' MSE: ' + str(mse))
    print(key + ' R2 ' + str(r2))
    print("Accuracy: %0.2f (+/- %0.2f)" % (scores.mean(), scores.std() * 2))
    if print_coef:
        print('Coefficients:')
        print(model.coef_)
    print('')

LinearRegression Train time: 0.006232051054636637 minutes
LinearRegression MSE: 1.8948411796124053e-26
LinearRegression R2 1.0
Accuracy: 1.00 (+/- 0.00)

Ridge Train time: 0.0008038640022277832 minutes
Ridge MSE: 3.762872372671515e-06
Ridge R2 0.9999999997330863
Accuracy: 1.00 (+/- 0.00)

Lasso Train time: 0.0001273194948832194 minutes
Lasso MSE: 1.0174272885629094
Lasso R2 0.999927830322463
Accuracy: 1.00 (+/- 0.00)

ElasticNet Train time: 0.00017528533935546876 minutes
ElasticNet MSE: 1598.1075012508559
ElasticNet R2 0.8866406431878385
Accuracy: 0.89 (+/- 0.00)

Lars Train time: 0.0008799393971761067 minutes
Lars MSE: 6.252958768736269e-27
Lars R2 1.0
Accuracy: 1.00 (+/- 0.00)

OrthogonalMatchingPursuit Train time: 0.00018830299377441405 minutes
OrthogonalMatchingPursuit MSE: 2.684416235869073e-27
OrthogonalMatchingPursuit R2 1.0
Accuracy: 1.00 (+/- 0.00)

BayesianRidge Train time: 0.0003278613090515137 minutes
BayesianRidge MSE: 3.251774052517294e-26
BayesianRidge R2 1.0
Accuracy: 1

MemoryError: 

In [ ]:
import pickle
# save the model to disk
filename = 'unemployment.sav'
pickle.dump(best_Model, open(filename, 'wb'))